lib/src/presentation/views/sign_in/sign_in_viewmodel.dart

```diff
@@ -0,0 +1,57 @@
+import 'package:basic_authentication_flutter_challenge/injection.dart';
+import 'package:basic_authentication_flutter_challenge/src/services/notifier.dart';
+import 'package:flutter/material.dart';
+import 'package:stacked/stacked.dart';
+import 'package:basic_authentication_flutter_challenge/src/services/auth_service.dart';
+
+class SignInViewModel extends BaseViewModel {
+  bool isLoading = false;
+  final formKey = GlobalKey<FormState>();
+  final usernameController = TextEditingController();
+  final passwordController = TextEditingController();
+
+  Future login() async {
+    if (isFormValid()) {
+      _setLoading(true);
+      try {
+        await locator<AuthService>().login(
+          usernameController.text,
+          passwordController.text,
+        );
+      } catch (e) {
+        locator<Notifier>().show('Couldn\'t authenticate: ${e.toString()}');
+      }
+      _setLoading(false);
+    }
+  }
+
+  bool isFormValid() {
+    return formKey.currentState?.validate() ?? false;
+  }
+
+  String? usernameValidator(String? value) {
+    if (value is! String || value.isEmpty) {
+      return 'Username can\'t be empty';
+    }
+    return null;
+  }
+
+  String? passwordValidator(String? value) {
+    if (value is! String || value.isEmpty) {
+      return 'Password can\'t be empty';
+    }
+    return null;
+  }
```

Comment:
I believe the condition might be optimized to avoid one unneeded check. The method always expects `String` and will complain if you pass `non-string` type.

In this validation we have to understand two aspects, form value and validator value. Since you are using `TextEditingController()` the value will be what the user have typed or null. In null case, we can see there is a logic to override the value to be empty string.

![Screen Shot 1444-02-07 at 2 20 14 PM](https://user-images.githubusercontent.com/54819958/188268056-f42f49ee-9737-4c6a-a7a2-f683550d40b6.png)

![Screen Shot 1444-02-07 at 2 15 21 PM](https://user-images.githubusercontent.com/54819958/188268095-dfa9d21a-e7bd-4fd2-b00d-da2ca10058d6.png)


---
So, we need need to validate the type anymore and to make sure you might check the type at runtime and it will always by `String`.

```dart
  String? usernameValidator(String? value) {
    /// run time
    var runtimeType = value.runtimeType;
    
    if (value is! String || value.isEmpty) {
      return 'Username can\'t be empty';
    }
    return null;
  }
```

An optimized condition might be similar to the below:
```suggestion
  String? usernameValidator(String? value) {
    if (value!.isEmpty) {
      return 'Username can\'t be empty';
    }
    return null;
  }

  String? passwordValidator(String? value) {
    if (value!.isEmpty) {
      return 'Password can\'t be empty';
    }
    return null;
  }
```

Please let me know if you've considered edge cases that didn't know about.

Reply:
I think your code will break if Flutter Team decided to pass a nullable value in the future. 
I always try to avoid `!`

Reply:
As you know I am not Flutter guy. I put `!` just because the complier yelled at me. I would love to know why are you trying to avoid it?

Reply:
Dart is a null-safe language. The complier will always tell you if a value might be null. 

if a variable has a type of `String` that means it can never be null, therefore you don't need to write an `if` statement to check if it is null or not.  

if a variable has a type of `String?`   that means it could be a `String` or `null`, therefore the compiler will force you to write an `if` statement to check if it is null or not. 

There are cases when you define a variable as `String?` but at some point, you know that it will not be null, that's where you can use `!`, to tell the compiler that I'm 100% sure that this value won't be null.

Reply:
It a good practice to avoid `!` and do what the compiler tells you.